openprescribing/frontend/tests/price_per_unit/test_savings.py

```diff
@@ -0,0 +1,320 @@
+        # Pick an arbitrary month and practice
+        date = self.factory.months[-1][:10]
+        practice = self.factory.practices[-1]
+
+        result = get_total_savings_for_org(date, "practice", practice["code"])
+
+        # Calculate the expected savings the boring way
+        expected = sum_savings(
+            get_savings_for_practice(
+                date,
+                practice,
+                self.factory.prescribing,
+                self.substitution_sets,
+                CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["practice"] / 100,
+            )
+        )
+
+        self.assertEqual(round_floats(result), round_floats(expected))
+
+    def test_all_england_savings(self):
+        # Pick an arbitrary month
+        date = self.factory.months[0][:10]
+
+        # Temporarily lower the threshold so we get some savings
+        orig_value = CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["all_standard_practices"]
+        CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["all_standard_practices"] = 100
+
+        # Call the API to get the savings
+        response = self.client.get(
+            "/api/1.0/price_per_unit/",
+            {"format": "json", "aggregate": "True", "date": date},
+        )
+        results = json.loads(response.content.decode("utf8"))
+
+        # Calculate the expected savings the boring way
+        expected = get_savings_for_all_england(
+            date,
+            self.factory.prescribing,
+            self.substitution_sets,
+            CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["all_standard_practices"] / 100,
+        )
+
+        # Reset config value
+        CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["all_standard_practices"] = orig_value
+
+        # Make sure we've actually got some savings to compare
+        self.assertTrue(results)
+        self.assertEqual(round_floats(results), round_floats(expected))
+
+    def test_all_england_total_savings(self):
+        # Pick an arbitrary month
+        date = self.factory.months[-1][:10]
+
+        # Temporarily lower the threshold so we get some savings
+        orig_value = CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["all_standard_practices"]
+        CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["all_standard_practices"] = 100
+
+        result = get_total_savings_for_org(date, "all_standard_practices", None)
+
+        # Calculate the expected savings the boring way
+        expected = sum_savings(
+            get_savings_for_all_england(
+                date,
+                self.factory.prescribing,
+                self.substitution_sets,
+                CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["all_standard_practices"] / 100,
+            )
+        )
+
+        # Reset config value
+        CONFIG_MIN_SAVINGS_FOR_ORG_TYPE["all_standard_practices"] = orig_value
+
+        self.assertEqual(round_floats(result), round_floats(expected))
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._remove_patch()
+        super().tearDownClass()
+
+
+def invent_generic_bnf_code(index):
```

Comment:
Does this live in a data factory?

Reply:
Probably. But then it should really use `next_id()` rather than requiring an index passed in manually. And then it would need to deal with numbers larger than 9 and it didn't feel worth it until we had an actual need to reuse it.